### DISCUSSION TITLE
replace string in html by variable in my server

### DISCUSSION THREAD

**Discussion by dingo35** [cesanta/mongoose#2651] [Ask questions, get answers] [answered]
Hello,

I am working on Arduino Core (ESP32) platform, and replacing old AsyncWebServer by mongoose webserver. All going well, great code!

But I struggle with one thing:
The current solution serves static  index.html file and before serving, replaces certain strings by variable values in the server code:
`
 
          String processor(const String& var){
          if (var == "APhostname") return APhostname;
` 

In index.html:
`                          <h1 id="Hostname" class="h5 mb-0 text-white-800">%APhostname%</h1>
`

Now I have been looking into SSI pattern file, but it seems the echo function can only echo a few variables that are predefined in SSI.

Can I solve this by a certain ssi.html file, if so, what does it look like? If not, what would be a good solution?

**Comment by scaprile** [comment, marked as answer]
https://mongoose.ws/documentation/#struct-mg_http_serve_opts
https://mongoose.ws/documentation/#mg_http_serve_dir

Please follow the guidelines in the [documentation](https://mongoose.ws/documentation/),  [examples](https://github.com/cesanta/mongoose/tree/master/examples/) and [tutorials](https://mongoose.ws/documentation/#tutorials-and-examples) available.

**Comment by dingo35** [comment]
Well thank you for your quick but not very helpfull answer.

Of course I know how to "activate" the ssi pattern file, that part is described very well.

But what that ssi.html file does look like, and if it will replace any internal variable, remains a mystery.

Do a grep -R ssi_patten and you will see there is _not one_ example on this....

**Reply by scaprile** [reply]
As you can see, there is little to document, Mongoose just supports the `include`tag (besides the fact that SSI and HTML are not specific to Mongoose):
https://github.com/cesanta/mongoose/blob/15bd8b4fe9592a12d29e0ca9fe1c5db3ff3f7263/mongoose.c#L7882-L7905

Mongoose users write C code to serve (for example) JSON responses or dynamically generate HTML content on the fly using (for example) https://mongoose.ws/documentation/#mg_http_reply or https://mongoose.ws/documentation/#mg_printf-mg_vprintf or several other functions described in our docs and exemplified in our tutorials; like https://mongoose.ws/documentation/tutorials/http-server/

That is, you either write AJAX-like code using JS to query the host for your content and generate the page on the client side, or you generate the page on the server before serving it to the client.
You have examples for both in the Web-UI section of our tutorials and in several of our examples, and their respective tutorials.
https://mongoose.ws/documentation/#web-ui

**Comment by dingo35** [comment]
Oh, and ChatGPT even states that mongoose wevlbserver does not have a SSI mechanism. Perhaps because it is only fed until 2021?
